c10/util/safe_numerics.h

```diff
@@ -31,6 +31,15 @@ C10_ALWAYS_INLINE bool add_overflows(uint64_t a, uint64_t b, uint64_t* out) {
 #endif
 }
 
+C10_ALWAYS_INLINE bool add_overflows(int64_t a, int64_t b, int64_t* out) {
+#if C10_HAS_BUILTIN_OVERFLOW()
+  return __builtin_add_overflow(a, b, out);
+#else
+  *out = a + b;
+  return *out < a;
```

Comment:
Hmm, `add_overflows` return true in case of overflow, isn't it? But if `b` is less than zero, then `*out` is supposed to be less than `a` isn't it?
```suggestion
  return b >= 0 ? *out < a : *out > a;
```